File: sdks/ts/memory/src/ingest/migrate-hash.test.ts
// SPDX-License-Identifier: Apache-2.0

import { describe, expect, it } from 'vitest'
import { createMemStore, toPath } from '../store/index.js'
import { hashSlug, hashSlugSHA256 } from './hash.js'
import { createHashMigrator, resolveHash } from './migrate-hash.js'

const RAW_DOCUMENTS_PREFIX = 'raw/documents'

const rawDocPath = (slug: string) => toPath(`${RAW_DOCUMENTS_PREFIX}/${slug}.md`)

describe('createHashMigrator', () => {
  it('migrates empty store with 0 migrated', async () => {
    const store = createMemStore()
    const migrator = createHashMigrator(store)

    const result = await migrator.migrate()
    expect(result.migrated).toBe(0)
    expect(result.total).toBe(0)
    expect(result.nextCursor).toBe('')
  })

  it('migrates documents from SHA-256 to BLAKE3 hashes', async () => {
    const store = createMemStore()

    const docs = [
      Buffer.from('document one content', 'utf8'),
      Buffer.from('document two content', 'utf8'),
      Buffer.from('document three content', 'utf8'),
    ]

    for (const content of docs) {
      const sha256Slug = hashSlugSHA256(content)
      await store.write(rawDocPath(sha256Slug), content)
    }

    const migrator = createHashMigrator(store)
    const result = await migrator.migrate({ batchSize: 10 })

    expect(result.migrated).toBe(3)
    expect(result.total).toBe(3)

    for (const content of docs) {
      const blake3SlugValue = hashSlug(content)
      const newExists = await store.exists(rawDocPath(blake3SlugValue))
      expect(newExists).toBe(true)

      const sha256SlugValue = hashSlugSHA256(content)
      const oldExists = await store.exists(rawDocPath(sha256SlugValue))
      expect(oldExists).toBe(false)
    }
  })

  it('resumes migration from saved cursor', async () => {
    const store = createMemStore()

    const docs = [
      Buffer.from('alpha content', 'utf8'),
      Buffer.from('beta content', 'utf8'),
      Buffer.from('gamma content', 'utf8'),
      Buffer.from('delta content', 'utf8'),
      Buffer.from('epsilon content', 'utf8'),
    ]

    for (const content of docs) {
      const sha256Slug = hashSlugSHA256(content)
      await store.write(rawDocPath(sha256Slug), content)
    }

    const migrator = createHashMigrator(store)

    const result1 = await migrator.migrate({ batchSize: 2 })
    expect(result1.migrated).toBe(2)
    expect(result1.total).toBe(5)
    expect(result1.nextCursor).not.toBe('')

    // Second pass resumes from saved state.
    const result2 = await migrator.migrate({ batchSize: 10 })
    const totalProcessed = result2.migrated + result2.skipped
    expect(totalProcessed).toBeGreaterThan(0)
  })

  it('dry-run reports but does not write', async () => {
    const store = createMemStore()

    const content = Buffer.from('dry run test content', 'utf8')
    const sha256Slug = hashSlugSHA256(content)
    const originalPath = rawDocPath(sha256Slug)
    await store.write(originalPath, content)

    const migrator = createHashMigrator(store)
    const result = await migrator.migrate({ dryRun: true })

    expect(result.migrated).toBe(1)

    // Original still exists.
    const originalExists = await store.exists(originalPath)
    expect(originalExists).toBe(true)

    // BLAKE3 path not created.
    const blake3SlugValue = hashSlug(content)
    const newExists = await store.exists(rawDocPath(blake3SlugValue))
    expect(newExists).toBe(false)

    // State file not written.
    const stateExists = await store.exists(toPath('raw/.pipeline-state/.blake3-migration.json'))
    expect(stateExists).toBe(false)
  })
})

describe('resolveHash', () => {
  it('finds document by BLAKE3 hash first', async () => {
    const store = createMemStore()

    const content = Buffer.from('dual-read test content', 'utf8')
    const blake3SlugValue = hashSlug(content)
    const sha256SlugValue = hashSlugSHA256(content)

    // Write at both paths.
    await store.write(rawDocPath(blake3SlugValue), content)
    await store.write(rawDocPath(sha256SlugValue), content)

    const resolved = await resolveHash(store, content)
    expect(resolved).toBe(blake3SlugValue)
  })

  it('falls back to SHA-256 when BLAKE3 not found', async () => {
    const store = createMemStore()

    const content = Buffer.from('legacy content sha256 only', 'utf8')
    const sha256SlugValue = hashSlugSHA256(content)
    await store.write(rawDocPath(sha256SlugValue), content)

    const resolved = await resolveHash(store, content)
    expect(resolved).toBe(sha256SlugValue)
  })

  it('returns BLAKE3 hash for new documents', async () => {
    const store = createMemStore()

    const content = Buffer.from('brand new document', 'utf8')
    const blake3SlugValue = hashSlug(content)

    const resolved = await resolveHash(store, content)
    expect(resolved).toBe(blake3SlugValue)
  })
})

File: go/ingest/migrate_hash_test.go
// SPDX-License-Identifier: Apache-2.0
package ingest

import (
	"context"
	"testing"

	"github.com/jeffs-brain/memory/go/brain"
	"github.com/jeffs-brain/memory/go/store/mem"
)

func TestHashMigrator_EmptyStore(t *testing.T) {
	t.Parallel()
	store := mem.New()
	migrator := NewHashMigrator(store)

	result, err := migrator.Migrate(context.Background(), MigrateOpts{})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result.Migrated != 0 {
		t.Errorf("expected 0 migrated, got %d", result.Migrated)
	}
	if result.Total != 0 {
		t.Errorf("expected 0 total, got %d", result.Total)
	}
	if result.NextCursor != "" {
		t.Errorf("expected empty cursor, got %q", result.NextCursor)
	}
}

func TestHashMigrator_MigratesDocuments(t *testing.T) {
	t.Parallel()
	ctx := context.Background()
	store := mem.New()

	// Write documents with SHA-256-based slugs.
	docs := []struct {
		content []byte
	}{
		{content: []byte("document one content")},
		{content: []byte("document two content")},
		{content: []byte("document three content")},
	}

	for _, doc := range docs {
		sha256Slug := HashSlugSHA256(doc.content)
		docPath := brain.RawDocument(sha256Slug)
		if err := store.Write(ctx, docPath, doc.content); err != nil {
			t.Fatalf("write failed: %v", err)
		}
	}

	migrator := NewHashMigrator(store)
	result, err := migrator.Migrate(ctx, MigrateOpts{BatchSize: 10})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result.Migrated != 3 {
		t.Errorf("expected 3 migrated, got %d", result.Migrated)
	}
	if result.Total != 3 {
		t.Errorf("expected 3 total, got %d", result.Total)
	}

	// Verify new paths exist with BLAKE3 slugs.
	for _, doc := range docs {
		blake3Slug := HashSlug(doc.content)
		newPath := brain.RawDocument(blake3Slug)
		exists, err := store.Exists(ctx, newPath)
		if err != nil {
			t.Fatalf("exists check failed: %v", err)
		}
		if !exists {
			t.Errorf("expected document at BLAKE3 path %s to exist", newPath)
		}

		// Verify old SHA-256 paths removed.
		sha256Slug := HashSlugSHA256(doc.content)
		oldPath := brain.RawDocument(sha256Slug)
		exists, err = store.Exists(ctx, oldPath)
		if err != nil {
			t.Fatalf("exists check failed: %v", err)
		}
		if exists {
			t.Errorf("expected old SHA-256 path %s to be removed", oldPath)
		}
	}
}

func TestResolveHash_FindsByBLAKE3First(t *testing.T) {
	t.Parallel()
	ctx := context.Background()
	store := mem.New()

	content := []byte("test content for dual-read")
	blake3Slug := HashSlug(content)
	blake3Path := brain.RawDocument(blake3Slug)
	if err := store.Write(ctx, blake3Path, content); err != nil {
		t.Fatalf("write failed: %v", err)
	}

	// Also write at SHA-256 path (simulating pre-migration state).
	sha256Slug := HashSlugSHA256(content)
	sha256Path := brain.RawDocument(sha256Slug)
	if err := store.Write(ctx, sha256Path, content); err != nil {
		t.Fatalf("write failed: %v", err)
	}

	resolved, err := ResolveHash(ctx, store, content)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if resolved != blake3Slug {
		t.Errorf("expected BLAKE3 slug %q, got %q", blake3Slug, resolved)
	}
}

func TestResolveHash_FallsBackToSHA256(t *testing.T) {
	t.Parallel()
	ctx := context.Background()
	store := mem.New()

	content := []byte("legacy content only at sha256 path")
	sha256Slug := HashSlugSHA256(content)
	sha256Path := brain.RawDocument(sha256Slug)
	if err := store.Write(ctx, sha256Path, content); err != nil {
		t.Fatalf("write failed: %v", err)
	}

	resolved, err := ResolveHash(ctx, store, content)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if resolved != sha256Slug {
		t.Errorf("expected SHA-256 slug %q, got %q", sha256Slug, resolved)
	}
}

func TestResolveHash_ReturnsBLAKE3ForNewDocuments(t *testing.T) {
	t.Parallel()
	ctx := context.Background()
	store := mem.New()

	content := []byte("brand new document never seen before")
	blake3Slug := HashSlug(content)

	resolved, err := ResolveHash(ctx, store, content)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if resolved != blake3Slug {
		t.Errorf("expected BLAKE3 slug %q for new document, got %q", blake3Slug, resolved)
	}
}

func TestHashMigrator_ResumeFromCursor(t *testing.T) {
	t.Parallel()
	ctx := context.Background()
	store := mem.New()

	// Write 5 documents with SHA-256 slugs.
	contents := [][]byte{
		[]byte("alpha content"),
		[]byte("beta content"),
		[]byte("gamma content"),
		[]byte("delta content"),
		[]byte("epsilon content"),
	}

	for _, content := range contents {
		sha256Slug := HashSlugSHA256(content)
		docPath := brain.RawDocument(sha256Slug)
		if err := store.Write(ctx, docPath, content); err != nil {
			t.Fatalf("write failed: %v", err)
		}
	}

	migrator := NewHashMigrator(store)

	// First pass: migrate only 2 documents.
	result1, err := migrator.Migrate(ctx, MigrateOpts{BatchSize: 2})
	if err != nil {
		t.Fatalf("first pass error: %v", err)
	}
	if result1.Migrated != 2 {
		t.Errorf("first pass: expected 2 migrated, got %d", result1.Migrated)
	}
	if result1.NextCursor == "" {
		t.Fatal("first pass: expected non-empty next cursor")
	}
	if result1.Total != 5 {
		t.Errorf("first pass: expected 5 total, got %d", result1.Total)
	}

	// Second pass: resume from cursor. The store has 5 total entries now
	// (2 new BLAKE3 + 3 remaining SHA-256). The remaining 3 SHA-256 docs
	// should be migrated.
	result2, err := migrator.Migrate(ctx, MigrateOpts{BatchSize: 10})
	if err != nil {
		t.Fatalf("second pass error: %v", err)
	}
	// Should migrate the remaining documents (some may show as skipped if
	// they were already migrated, depends on cursor position in the new
	// listing).
	totalProcessed := result2.Migrated + result2.Skipped
	if totalProcessed == 0 {
		t.Error("second pass: expected some documents to be processed")
	}
}

func TestHashMigrator_DryRunDoesNotWrite(t *testing.T) {
	t.Parallel()
	ctx := context.Background()
	store := mem.New()

	content := []byte("dry run test content")
	sha256Slug := HashSlugSHA256(content)
	docPath := brain.RawDocument(sha256Slug)
	if err := store.Write(ctx, docPath, content); err != nil {
		t.Fatalf("write failed: %v", err)
	}

	migrator := NewHashMigrator(store)
	result, err := migrator.Migrate(ctx, MigrateOpts{DryRun: true})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result.Migrated != 1 {
		t.Errorf("expected 1 migrated (reported), got %d", result.Migrated)
	}

	// Verify original file still exists (dry-run did not modify).
	exists, err := store.Exists(ctx, docPath)
	if err != nil {
		t.Fatalf("exists check failed: %v", err)
	}
	if !exists {
		t.Error("dry-run should not have removed the original document")
	}

	// Verify BLAKE3 path was NOT created.
	blake3Slug := HashSlug(content)
	newPath := brain.RawDocument(blake3Slug)
	exists, err = store.Exists(ctx, newPath)
	if err != nil {
		t.Fatalf("exists check failed: %v", err)
	}
	if exists {
		t.Error("dry-run should not have created the BLAKE3 path")
	}

	// Verify no state file written.
	statePath := brain.Path(migrationStatePath)
	exists, err = store.Exists(ctx, statePath)
	if err != nil {
		t.Fatalf("exists check failed: %v", err)
	}
	if exists {
		t.Error("dry-run should not have written migration state")
	}
}
